packages/cache/src/cache.ts

```diff
@@ -0,0 +1,175 @@
+  if (key.length > 512) {
+    throw new ValidationError(
+      `Key Validation Error: ${key} cannot be larger than 512 characters.`
+    )
+  }
+  const regex = /^[^,]*$/
+  if (!regex.test(key)) {
+    throw new ValidationError(
+      `Key Validation Error: ${key} cannot contain commas.`
+    )
+  }
+}
+
+/**
+ * Restores cache from keys
+ *
+ * @param paths a list of file paths to restore from the cache
+ * @param primaryKey an explicit key for restoring the cache
+ * @param restoreKeys an optional ordered list of keys to use for restoring the cache if no cache hit occurred for key
+ * @returns string returns the key for the cache hit, otherwise return undefined
+ */
+export async function restoreCache(
+  paths: string[],
+  primaryKey: string,
+  restoreKeys?: string[]
+): Promise<string | undefined> {
+  checkPaths(paths)
+
+  restoreKeys = restoreKeys || []
+  const keys = [primaryKey, ...restoreKeys]
+
+  core.debug('Resolved Keys:')
+  core.debug(JSON.stringify(keys))
+
+  if (keys.length > 10) {
+    throw new ValidationError(
+      `Key Validation Error: Keys are limited to a maximum of 10.`
+    )
+  }
+  for (const key of keys) {
+    checkKey(key)
+  }
+
+  const compressionMethod = await utils.getCompressionMethod()
+
+  // path are needed to compute version
+  const cacheEntry = await cacheHttpClient.getCacheEntry(keys, paths, {
+    compressionMethod
+  })
+  if (!cacheEntry?.archiveLocation) {
+    core.info(`Cache not found for input keys: ${keys.join(', ')}`)
+    return undefined
+  }
+
+  const archivePath = path.join(
+    await utils.createTempDirectory(),
+    utils.getCacheFileName(compressionMethod)
+  )
+  core.debug(`Archive Path: ${archivePath}`)
+
+  try {
+    // Download the cache from the cache entry
+    await cacheHttpClient.downloadCache(cacheEntry.archiveLocation, archivePath)
+
+    const archiveFileSize = utils.getArchiveFileSize(archivePath)
+    core.info(
+      `Cache Size: ~${Math.round(
+        archiveFileSize / (1024 * 1024)
+      )} MB (${archiveFileSize} B)`
+    )
+
+    await extractTar(archivePath, compressionMethod)
+  } finally {
+    // Try to delete the archive to save space
+    try {
+      await utils.unlinkFile(archivePath)
+    } catch (error) {
+      core.debug(`Failed to delete archive: ${error}`)
+    }
+  }
+
+  core.info(`Cache restored from key: ${cacheEntry && cacheEntry.cacheKey}`)
+
+  return cacheEntry.cacheKey
+}
+
+/**
+ * Saves a list of files with the specified key
+ *
+ * @param paths a list of file paths to be cached
+ * @param key an explicit key for restoring the cache
+ * @param options cache upload options
+ * @returns number returns cacheId if the cache was saved successfully
+ */
+export async function saveCache(
+  paths: string[],
+  key: string,
+  options?: UploadOptions
+): Promise<number> {
+  checkPaths(paths)
+  checkKey(key)
+
+  const compressionMethod = await utils.getCompressionMethod()
+
+  core.debug('Reserving Cache')
```

Comment:
Just an observation, `core.debug` is used pretty frequently, but I think there should be some uses of `core.info` so users see more logs and information.

Most users won't have step-debugging turned on and I've noticed that a lot of our first party actions don't have the greatest output/logging.

Reply:
Package users/Action Authors might not want that verbose logging from a package they consume. I'd say leave as-is and if we get feedback we can add an option to the package to add verbose logging rather than verbose by default

Reply:
Fine for now, just be aware that once `actions/cache` switches over to using this package, the same level of logging will apply (unless there is an option)